### PR TITLE
feat: Bandits part 3: Bandit model DTOs

### PR DESCRIPTION
**Pull request by typotter**
### Motivation and context
In order to compute bandit actions, the model data needs to be loaded and parsed. This simple PR introduces the data types needed for Bandit evaluation with built-in parsing from decoded JSON.

### Description
- DTOs for Bandit models and evaluation/log event.
- Parsing built into DTOs
- DTOs for passing user data into the Client (subject and action attributes) as well as convenience methods to make passing data into the client more flexible for the end-user (developers)

### How is this tested
- Unit tests for the helper parsing logic we get in `AttributeSet` and `ContextAttributes`
`BanditDTOTest` parses and verifies a whole tree of data.

### CODE REVIEW

File: src/Bandits/BanditVariationIndexer.php [hunk 1]
```diff
@@ -41,20 +41,19 @@ public function __construct(array $banditVariations)
...
-
     public function getBanditByVariation($flagKey, $variation): ?string
     {
         return $this->banditFlags[$flagKey][$variation] ?? null;
     }
 
     public function isBanditFlag($flagKey): bool
     {
-        return array_key_exists($flagKey, $this->banditFlags);
+        return isset($this->banditFlags[$flagKey]);
```
Comment (typotter) [inline src/Bandits/BanditVariationIndexer.php:57]: faster resolution than array_key_exists.

File: src/DTO/Bandit/ActionCoefficients.php [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+     */
+    public function __construct(
+        public readonly string $actionKey,
+        public readonly float $intercept,
+        public readonly array $subjectNumericCoefficients = [],
+        public readonly array $subjectCategoricalCoefficients = [],
+        public readonly array $actionNumericCoefficients = [],
+        public readonly array $actionCategoricalCoefficients = []
+    ) {
+        // Since PHP doesn't enforce array typing and this dev spent an unfortunate amount of time debugging, we ensure
```
Comment (typotter) [inline src/DTO/Bandit/ActionCoefficients.php:26]: higher levels of DTOs properly enforce the typings

File: src/DTO/Bandit/BanditEvaluationEvent.php [hunk 3]
```diff
@@ -0,0 +1,47 @@
+<?php
+
+namespace Eppo\DTO\Bandit;
+
+use DateTime;
+use JsonSerializable;
+
+class BanditEvaluationEvent implements JsonSerializable
```
Comment (typotter) [inline src/DTO/Bandit/BanditEvaluationEvent.php:8]: many of the telemetry packages accept a dictionary/map-like structure for userdata.

File: src/DTO/Bandit/ActionCoefficients.php (hunk 2 above)
Comment (aarsilv) [reply src/DTO/Bandit/ActionCoefficients.php:26]: I love how the comment alludes to this decision being made from blood sweat and tears 😅

File: src/DTO/Bandit/AttributeSet.php [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+     * @param array<string, float> $numericAttributes
+     * @param array<string, bool|float|int|string> $categoricalAttributes
+     */
+    public function __construct(
+        public readonly array $numericAttributes = [],
+        public readonly array $categoricalAttributes = []
+    ) {
+    }
+
+    public static function fromArray(array $attributes): self
```
Comment (aarsilv) [inline src/DTO/Bandit/AttributeSet.php:31]: ❤️

File: src/DTO/Bandit/AttributeSet.php [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+{
+    /**
+     * Types to automatically classify as Categorical
+     */
+    private const CATEGORICAL_TYPES = ["boolean", "string"];
+
+    /**
+     * Types to automatically classify as Numeric
+     * `float` and `double` are the sane datatype in php.
+     * In PHP, for historical reasons, `gettype` returns `double`.
```
Comment (aarsilv) [inline src/DTO/Bandit/AttributeSet.php:15]: I appreciate the comment--I was just wondering if the two were different somewhow!

File: src/DTO/Bandit/IContextAttributes.php [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+namespace Eppo\DTO\Bandit;
+
+/**
+ * A set of attributes for a given context ($key))
+ */
+interface IContextAttributes
+{
+    public function getAttributes(): AttributeSet;
+
+    public function getKey(): string;
```
Comment (aarsilv) [inline src/DTO/Bandit/IContextAttributes.php:12]: What do we need `$key` for? Each attribute should have it's own key.

File: tests/DTO/Bandit/AttributeSetTest.php [hunk 7]
```diff
@@ -0,0 +1,70 @@
+<?php
+
+namespace Eppo\Tests\DTO\Bandit;
+
+use Eppo\DTO\Bandit\AttributeSet;
+use PHPUnit\Framework\TestCase;
+
+class AttributeSetTest extends TestCase
```
Comment (aarsilv) [inline tests/DTO/Bandit/AttributeSetTest.php:8]: 🙌

File: tests/DTO/Bandit/AttributeSetTest.php [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+            'city' => 'London',
+            'subscribed' => false,
+            'areaCode' => "90210"
+        );
+
+        $this->assertEquals($expectedNumericAttributes, $attributeSet->numericAttributes);
+        $this->assertEquals($expectedCategoricalAttributes, $attributeSet->categoricalAttributes);
+    }
+
+    public function testFromArrayFiltersInvalidValueTypes()
```
Comment (aarsilv) [inline tests/DTO/Bandit/AttributeSetTest.php:49]: 🔥

File: tests/DTO/BanditDTOTest.php [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+    }';
+
+    public function testParsesFromJson(): void
+    {
+        $json = json_decode(self::BANDIT_JSON, true);
+        $bandit = Bandit::fromJson($json);
+
+        $this->assertNotNull($bandit);
+
+        // Assert on most properties down the entire object tree, testing the fromJSON methods of each DTO.
```
Comment (aarsilv) [inline tests/DTO/BanditDTOTest.php:120]: 💪

**Review by aarsilv — APPROVED** [review]
Looking good! My only substantive comment is that I don't think `ContextAttributes` needs its own key given all of the attributes nested within have their keys.

File: src/DTO/Bandit/IContextAttributes.php (hunk 6 above)
Comment (typotter) [reply src/DTO/Bandit/IContextAttributes.php:19]: `ContextAttributes` pairs a string `key` with an `AttributeSet` in order to keep SubjectIDs and ActionKeys closer to their attributes when dealing with groups of same. It's a convenience layer that alleviates some awkwardness with having to access both key and value when manipulating arrays.

I've updated the docs on what exactly the `IContextAttributes` interface is intended for/what problem it solves. Hopefully that makes it clear why the added layer on top of `AttributeSet`

**Comment by typotter** [comment]
Thanks @aarsilv . I addressed your comments on `ContextAttributes::$key` and added filtering in `AttributeSet` to pre-emptively filter out non-numerics.

**Pull request merged**